src/addresses.rs

```diff
@@ -0,0 +1,198 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::{collections::HashSet, sync::Arc};
+
+use multiaddr::{Multiaddr, Protocol};
+use parking_lot::RwLock;
+
+use crate::PeerId;
+
+/// Set of the public addresses of the local node.
+///
+/// The format of the addresses stored in the set contain the local peer ID.
+/// This requirement is enforced by the [`PublicAddresses::add_address`] method,
+/// that will add the local peer ID to the address if it is missing.
+///
+/// # Note
+///
+/// - The addresses are reported to the identify protocol and are used by other nodes
+///  to establish a connection with the local node.
+///
+/// - Users must ensure that the addresses are reachable from the network.
+#[derive(Debug, Clone)]
+pub struct PublicAddresses {
+    pub(crate) inner: Arc<RwLock<HashSet<Multiaddr>>>,
+    local_peer_id: PeerId,
+}
+
+impl PublicAddresses {
+    /// Creates new [`PublicAddresses`] from the given peer ID.
+    pub(crate) fn new(local_peer_id: PeerId) -> Self {
+        Self {
+            inner: Arc::new(RwLock::new(HashSet::new())),
+            local_peer_id,
+        }
+    }
+
+    /// Add a public address to the list of addresses.
+    ///
+    /// The address must contain the local peer ID, otherwise an error is returned.
+    /// In case the address does not contain any peer ID, it will be added.
+    ///
+    /// Returns true if the address was added, false if it was already present.
+    pub fn add_address(&self, address: Multiaddr) -> Result<bool, InsertionError> {
+        let address = ensure_local_peer(address, self.local_peer_id)?;
+        Ok(self.inner.write().insert(address))
+    }
+
+    /// Remove the exact public address.
+    ///
+    /// The provided address must contain the local peer ID.
+    pub fn remove_address(&self, address: &Multiaddr) -> bool {
+        self.inner.write().remove(address)
+    }
+
+    /// Returns a vector of the available listen addresses.
+    pub fn get_addresses(&self) -> Vec<Multiaddr> {
+        self.inner.read().iter().cloned().collect()
+    }
+}
+
+/// Set of the addresses the local node listens on.
+///
+/// The format of the addresses stored in the set contain the local peer ID.
+/// This requirement is enforced by the [`ListenAddresses::add_address`] method.
+///
+/// The listen addresses are populated during the construction of the Litep2p object.
+#[derive(Debug, Clone)]
+pub struct ListenAddresses {
```

Comment:
Actually, do we need a different struct? May be we can go with just one `Addresses`, and instantiate it to manage `listen_addresses` and `public_addresses`?

One thing with `ListenAddresses` though, is that ideally we want to start listening on the newly added address there. So, I'm not sure if we need to provide API for this struct, or just completely keep the old implementation of `listen_addresses` and use a dedicated struct only for `PublicAddresses`.

Reply:
That makes sense. Yep ideally we should be able to say `litep2p.register_listen_addr` and then give that information to individual protocol listeners. In this case, I'll just keep public addresses for this PR since a `ListenAddr` struct would complicate that implementation, thanks!